rustls/src/dns_name.rs

```diff
@@ -0,0 +1,176 @@
+/// DNS name validation according to RFC1035, but with underscores allowed.
+use std::error::Error as StdError;
+use std::fmt;
+
+/// "Labels must be 63 characters or less."
+const MAX_LABEL_LENGTH: usize = 63;
+
+fn validate(input: &[u8]) -> Result<(), InvalidDnsNameError> {
+    let mut label_length = 0;
+    let mut previous_label_ch = b'-';
+
+    for ch in input {
+        match ch {
+            // period is illegal other than after a non-empty label
+            b'.' if label_length == 0 => {
+                return Err(InvalidDnsNameError);
+            }
+            // period is otherwise legal
+            b'.' => {
+                // but the last character of the previous label must not be a hyphen
+                if previous_label_ch == b'-' {
+                    return Err(InvalidDnsNameError);
+                }
+                label_length = 0;
+            }
+            // first character of label must be <letter>
+            b'a'..=b'z' | b'A'..=b'Z' if label_length == 0 => {
+                previous_label_ch = *ch;
+                label_length += 1;
+            }
+            // subsequent characters can be <letter> | <digit> | "-" | "_"
+            // (underscore is a relaxation of RFC1035)
+            b'a'..=b'z' | b'A'..=b'Z' | b'-' | b'_' | b'0'..=b'9' if label_length > 0 => {
+                previous_label_ch = *ch;
+                label_length += 1;
+            }
+            // other characters are illegal
+            _ => {
+                return Err(InvalidDnsNameError);
+            }
+        };
+
+        if label_length > MAX_LABEL_LENGTH {
+            return Err(InvalidDnsNameError);
+        }
+    }
+
+    if previous_label_ch == b'-' {
+        return Err(InvalidDnsNameError);
+    }
+
+    Ok(())
+}
+
+/// The provided input could not be parsed because
+/// it is not a syntactically-valid DNS Name.
+#[derive(Debug)]
+pub struct InvalidDnsNameError;
+
+impl fmt::Display for InvalidDnsNameError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("invalid dns name")
+    }
+}
+
+impl StdError for InvalidDnsNameError {}
+
+/// A type which encapsulates an owned string that is a syntactically valid DNS name.
+#[derive(Clone, Eq, Hash, PartialEq, Debug)]
+pub struct DnsName(String);
+
+impl<'a> DnsName {
+    /// Produce a borrowed `DnsNameRef` from this owned `DnsName`.
+    pub fn borrow(&'a self) -> DnsNameRef<'a> {
```

Comment:
Maybe this should implement the `Borrow` trait, and we could implement the `ToOwned` trait on `DnsNameRef`?

Reply:
I tried that initially and didn't get it to work. `Borrow` wants to return a `&DnsNameRef<'a>` which would be a temporary. That would be OK, except the callers want a `DnsNameRef<'a>` and making `DnsNameRef` `Copy + Clone` broke `ToOwned`:

```
error[E0119]: conflicting implementations of trait `ToOwned` for type `dns_name::DnsNameRef<'_>`
  --> rustls/src/dns_name.rs:55:1
   |
55 | impl<'a> ToOwned for DnsNameRef<'a> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `alloc`:
           - impl<T> ToOwned for T
             where T: Clone;
```

The full diff is:

```
diff --git a/rustls/src/client/client_conn.rs b/rustls/src/client/client_conn.rs
index f77b5508..f5f8a53a 100644
--- a/rustls/src/client/client_conn.rs
+++ b/rustls/src/client/client_conn.rs
@@ -25,6 +25,7 @@ use std::marker::PhantomData;
 use std::net::IpAddr;
 use std::ops::{Deref, DerefMut};
 use std::sync::Arc;
+use std::borrow::Borrow;
 use std::{fmt, io, mem};
 
 /// A trait for the ability to store client session data, so that sessions
@@ -367,7 +368,7 @@ impl ServerName {
     /// in the handshake.
     pub(crate) fn for_sni(&self) -> Option<dns_name::DnsNameRef> {
         match self {
-            Self::DnsName(dns_name) => Some(dns_name.borrow()),
+            Self::DnsName(dns_name) => Some(*dns_name.borrow()),
             Self::IpAddress(_) => None,
         }
     }
diff --git a/rustls/src/dns_name.rs b/rustls/src/dns_name.rs
index dde14502..18b6d0e9 100644
--- a/rustls/src/dns_name.rs
+++ b/rustls/src/dns_name.rs
@@ -1,5 +1,6 @@
 /// DNS name validation according to RFC1035, but with underscores allowed.
 use std::error::Error as StdError;
+use std::borrow::Borrow;
 use std::fmt;
 
 /// A type which encapsulates an owned string that is a syntactically valid DNS name.
@@ -7,11 +8,6 @@ use std::fmt;
 pub struct DnsName(String);
 
 impl<'a> DnsName {
-    /// Produce a borrowed `DnsNameRef` from this owned `DnsName`.
-    pub fn borrow(&'a self) -> DnsNameRef<'a> {
-        DnsNameRef(self.as_ref())
-    }
-
     /// Validate the given bytes are a DNS name if they are viewed as ASCII.
     pub fn try_from_ascii(bytes: &[u8]) -> Result<Self, InvalidDnsNameError> {
         // nb. a sequence of bytes that is accepted by `validate()` is both
@@ -22,6 +18,13 @@ impl<'a> DnsName {
     }
 }
 
+impl<'a> Borrow<DnsNameRef<'a>> for DnsName {
+    /// Produce a borrowed `DnsNameRef` from this owned `DnsName`.
+    fn borrow(&self) -> &'a DnsNameRef<'a> {
+        &DnsNameRef(self.as_ref())
+    }
+}
+
 impl TryFrom<String> for DnsName {
     type Error = InvalidDnsNameError;
 
@@ -38,15 +41,10 @@ impl AsRef<str> for DnsName {
 }
 
 /// A type which encapsulates a borrowed string that is a syntactically valid DNS name.
-#[derive(Eq, Hash, PartialEq, Debug)]
+#[derive(Eq, Hash, PartialEq, Debug, Copy, Clone)]
 pub struct DnsNameRef<'a>(&'a str);
 
 impl<'a> DnsNameRef<'a> {
-    /// Copy this object to produce an owned `DnsName`.
-    pub fn to_owned(&'a self) -> DnsName {
-        DnsName(self.0.to_string())
-    }
-
     /// Copy this object to produce an owned `DnsName`, smashing the case to lowercase
     /// in one operation.
     pub fn to_lowercase_owned(&'a self) -> DnsName {
@@ -54,6 +52,15 @@ impl<'a> DnsNameRef<'a> {
     }
 }
 
+impl<'a> ToOwned for DnsNameRef<'a> {
+    type Owned = DnsName;
+
+    /// Copy this object to produce an owned `DnsName`.
+    fn to_owned(&self) -> Self::Owned {
+        DnsName(self.0.to_string())
+    }
+}
+
 impl<'a> TryFrom<&'a str> for DnsNameRef<'a> {
     type Error = InvalidDnsNameError;
 
diff --git a/rustls/src/msgs/handshake.rs b/rustls/src/msgs/handshake.rs
index f8be007e..07423c30 100644
--- a/rustls/src/msgs/handshake.rs
+++ b/rustls/src/msgs/handshake.rs
@@ -16,6 +16,7 @@ use crate::verify::DigitallySignedStruct;
 
 use std::collections;
 use std::fmt;
+use std::borrow::Borrow;
 
 /// Create a newtype wrapper around a given type.
 ///
@@ -295,7 +296,7 @@ impl ConvertServerNameList for [ServerName] {
     fn get_single_hostname(&self) -> Option<DnsNameRef> {
         fn only_dns_hostnames(name: &ServerName) -> Option<DnsNameRef> {
             if let ServerNamePayload::HostName(ref dns) = name.payload {
-                Some(dns.borrow())
+                Some(*dns.borrow())
             } else {
                 None
             }

```

Reply:
Fair enough!